// dedalus-labs/lib/runner/messages.js
"use strict";
// ==============================================================================
//                  © 2025 Dedalus Labs, Inc. and affiliates
//                            Licensed under MIT
//           github.com/dedalus-labs/dedalus-sdk-typescript/LICENSE
// ==============================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildInitialMessages = buildInitialMessages;
exports.coerceToolCall = coerceToolCall;
/** Constructs initial conversation history from input parameters. */
function buildInitialMessages({ instructions, input, messages, }) {
    if (instructions && messages?.some((m) => m.role === 'system')) {
        throw new Error("Cannot supply both 'instructions' and a system message in 'messages'.");
    }
    let conversation = [];
    if (messages) {
        conversation = [...messages];
    }
    else if (typeof input === 'string') {
        conversation = [{ role: 'user', content: input }];
    }
    else if (Array.isArray(input)) {
        conversation = [...input];
    }
    if (instructions) {
        conversation.unshift({ role: 'system', content: instructions });
    }
    if (!conversation.length) {
        throw new Error('Must supply one of instructions/messages/input');
    }
    return conversation;
}
/** Normalizes tool call to standard format with required fields. */
function coerceToolCall(call) {
    const fn = call.function ?? {};
    let args = fn.arguments ?? '{}';
    if (typeof args !== 'string') {
        args = JSON.stringify(args);
    }
    return {
        id: call.id ?? '',
        type: call.type ?? 'function',
        function: {
            name: fn.name ?? '',
            arguments: args,
        },
    };
}
